File: pp-exercises/openmp/master-worker-omp.f90
!==============================================================================
module taskinfo

  integer, save :: maxtask = 0
  integer, save :: task  = 0

end module taskinfo

!------------------------------------------------------------------------------
program main

  use taskinfo
  implicit none
  
  maxtask = 40

  !$omp parallel
  call work()
  !$omp end parallel

end

!------------------------------------------------------------------------------
subroutine get_task(nexttask)

  use taskinfo
  implicit none
  integer, intent(out) :: nexttask

  !$omp critical
  if (task < maxtask) then
     task = task + 1
     nexttask = task
  else
     nexttask = -1
  endif
  !$omp end critical

end

!------------------------------------------------------------------------------
subroutine work()

  implicit none
  integer :: task
  integer, external :: omp_get_thread_num

  do
     call get_task(task)

     if (task < 0) exit
     write(6,*) "thread ", omp_get_thread_num(), ": working on task ", task
     call system("sleep 1")
  enddo

end

!==============================================================================

File: pp-exercises/openmp/master-worker-omp.c
# include <stdio.h>
# include <stdlib.h>
# include <omp.h>

struct {
    int maxtask;
    int task;
} taskinfo = { 0, 0 };

void work(void);
void get_task(int *nexttask);

int main(int argc, char *argv[])
{
    taskinfo.maxtask = 40;

    #pragma omp parallel
    work();

    return 0;
}

void get_task(int *nexttask)
{
    #pragma omp critical
    {
	if (taskinfo.task < taskinfo.maxtask) {
	    ++taskinfo.task;
	    *nexttask = taskinfo.task;
	} else {
            *nexttask = -1;
	}
    }
}

void work(void)
{
    int task;

    do
    {
	get_task(&task);

	if (task >= 0)
	{
	    printf("thread %d: working on task %d\n", 
		   (int) omp_get_thread_num(), task);
	    system("sleep 1");
	}
    }
    while (task >= 0);
}
